api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto

```diff
@@ -119,4 +120,8 @@ message RateLimitPerRoute {
 
   // Specifies if the rate limit filter should include the virtual host rate limits.
   VhRateLimitsOptions vh_rate_limits = 1 [(validate.rules).enum = {defined_only: true}];
+
+  // Specifies a set of rate limit configurations that can be applied to typed filter config.
+  // [#not-implemented-hide:]
+  repeated config.route.v3.RateLimit rate_limits = 2;
```

Comment:
If I understand this correctly:

-  `config.route.v3.RouteAction.rate_limits` will be migrated to `config.filter.http.rate_limit.v3.RateLimitPerRoute.rate_limits`
- `config.route.v3.VirtualHost` will be migrated to `config.filter.http.rate_limit.v3.RateLimit.rate_limits`

So we should also add `repeated config.route.v3.RateLimit rate_limits = 2;` to `message RateLimit` above, right?

Reply:
`message RateLimit` above is for top-level(i.e. per-HCM ) filter config.  
`RateLimitPerRoute` here is for any lower levels filter config override which includes per-route, per-virtual_host or per-clusterweight. 

`config.route.v3.RateLimit` which is added here and currently sitting inside `config.route.v3.RouteAction` is for lower level config override. (e.g. [per_route] (https://github.com/envoyproxy/envoy/blob/ad63b1c385fe95ddd6d547a43583904c1f459abe/source/common/router/config_impl.cc#L354)  and [per-virtualHost](https://github.com/envoyproxy/envoy/blob/ad63b1c385fe95ddd6d547a43583904c1f459abe/source/common/router/config_impl.cc#L1224))
AFAIU, that is what we want to moved from route/VirtualHost to typed filter config.

Also, cced @kyessenov  and @mattklein123  who participated in the original discussion #14732

Reply:
> AFAIU, that is what we want to move from route/VirtualHost to typed filter config.

Yeah. I would really appreciate it if we could follow through with the deprecation and cleanup of the fields that live within the route table directly.

Reply:
I agree with @mattklein123.  I don't think this file should depend on route_components.proto.  Instead, I think we should just copy the existing `RateLimit` proto from route_components.proto into this file, and then deprecate the copy in route_components.proto.  (Although if the plan here is not to actually implement this in Envoy yet, then the deprecation of the existing fields should wait for that implementation.)  That way, if we ever release v4, we can remove the `RateLimit` proto from route_components.proto.

@tyxia is correct that this `RateLimitPerRoute` proto, despite its name, is actually a filter override proto, which can be specified in the typed_per_filter_config field at the [VirtualHost level](https://github.com/envoyproxy/envoy/blob/438f62305af99fb37ba737b2da19948d3e269c56/api/envoy/config/route/v3/route_components.proto#L145), [Route level](https://github.com/envoyproxy/envoy/blob/438f62305af99fb37ba737b2da19948d3e269c56/api/envoy/config/route/v3/route_components.proto#L264), or [ClusterWeight level](https://github.com/envoyproxy/envoy/blob/438f62305af99fb37ba737b2da19948d3e269c56/api/envoy/config/route/v3/route_components.proto#L403).

Note that individual filters decide what goes in the override configs at these levels and what the behavior is if their override config exists at multiple levels for a given request (e.g., both at the VirtualHost and at the Route level).  Some filters will simply use the most specific one and ignore the higher-up one (e.g., use the Route-level one and ignore the VirtualHost-level one).  Other filters will merge the data in some appropriate way.

I don't have all of the context on the rate limit filter specifically, but given the discussion in #14732, it seems like we need some way to merge the data in this case.  We should think through how we want this to work.  And I think we want the merge semantics to work at any of the three levels of filter override configs, not just between VirtualHost and Route -- e.g., the semantics should also be clear if the override config is present at the Route level and the ClusterWeight level but not the VirtualHost level.  I think this implies that the existing `VhRateLimitsOptions` field above needs to be generalized somehow.

Reply:
>  (Although if the plan here is not to actually implement this in Envoy yet, then the deprecation of the existing fields should wait for that implementation.)

IMO the implementation is not difficult. Basically in envoy we just take the deprecated field and create a synthetic per-filter config for it. I would really appreciate it we just do it per above.

Reply:
Matt, given the intended semantic change for the key names in `typed_per_filter_config` in #12274 (which gRPC has already implemented), what key would Envoy use for its automatically created per-filter config?  Would it just use the legacy well-known filter name for the ratelimit filter?  If so, then when Envoy actually implements the change for #12274, it won't be possible to use a filter instance name *other* than the legacy well-known filter name with the old fields.  That seems like it's probably a reasonable restriction, but I just want to verify that there aren't going to be any unexpected problems here.

Assuming that's not an issue, then if the implementation is really that simple, then I agree that we should go ahead and implement this.  @tyxia, what do you think?

Reply:
> That seems like it's probably a reasonable restriction, but I just want to verify that there aren't going to be any unexpected problems here.

IMO that's a fine restriction which we can document.

Reply:
SGTM.  Thanks!

Reply:
> Assuming that's not an issue, then if the implementation is really that simple, then I agree that we should go ahead and implement this. @tyxia, what do you think?

Sure. I can certainly work on that!  I just want to confirm the implementation/clean-up should be a separate follow-up PR and in this API PR we only do 1) Copy `message ratelimit` definition to the file here 2) Add `ratelimit` into `message RateLimitPerRoute` .

Reply:
I would rather just do it in one shot because then it will be more likely it will actually get done. ;)

Reply:
@mattklein123 are you thinking of doing both, implementing the use of the new API in Envoy as well as deprecating old API with associated changes to tests, docs, examples? We originally planned API only changed and the extra work would need to be reflected in internal schedule.

Reply:
Yes I would rather do both. The issue with doing a change like this is that we accrue debt and the likelihood of someone actually coming back and fixing it is pretty minimal.

Reply:
Hey @mattklein123. The reason why we're considering the phased approach here (proto first, then cleanup) is gRPC. We're adding RLS support, and would love to use the new fields from the beginning. Doing the cleanup at the same time is great, but jeopardizes the timeline of our Envoy folks. They want to be confident in delivering Quota for descriptor sets (#17954) and RLS streaming. These are to address reliability issues, and they're prioritized first.

So, unfortunately, now the choice is between not delivering this proto change (and enquiring tech debt on gRPC side, e.g. supporting old fields for about a year), or splitting this PR into two phases.

On the bright side, since gRPC will be using the new fields from the beginning, the cleanup on Envoy's side **would have** to happen. Even if gRPC team has to do it, it's still less cost than supporting old-style fields in three languages+all bindings. I agree with you about empty cleanup/tech debt promises, but, I hope, you can see this time it's not the case.

I appreciate your understanding in the matter.

Reply:
Fair enough, but someone please actually do the cleanup! Thank you.

Reply:
Great! Thank you so much for understanding.

Reply:
Created tracking issue for the second phase: https://github.com/envoyproxy/envoy/issues/18118

Reply:
Thanks all for discussion and insights! I resolved this conversation as #18118 is filed (Thanks, Sergii) and assigned.